src/backend/distributed/commands/alter_table.c

```diff
@@ -0,0 +1,1297 @@
+	{
+		CreateCitusTableLike(con);
+	}
+
+	ReplaceTable(con->relationId, con->newRelationId, justBeforeDropCommands);
+
+	TableDDLCommand *tableConstructionCommand = NULL;
+	foreach_ptr(tableConstructionCommand, postLoadCommands)
+	{
+		Assert(CitusIsA(tableConstructionCommand, TableDDLCommand));
+		char *tableConstructionSQL = GetTableDDLCommand(tableConstructionCommand);
+		ConnectAndExecuteSPIQuery(tableConstructionSQL, SPI_OK_UTILITY);
+	}
+
+	char *attachPartitionCommand = NULL;
+	foreach_ptr(attachPartitionCommand, attachPartitionCommands)
+	{
+		Node *parseTree = ParseTreeNode(attachPartitionCommand);
+
+		CitusProcessUtility(parseTree, attachPartitionCommand, PROCESS_UTILITY_TOPLEVEL,
+							NULL, None_Receiver, NULL);
+	}
+
+	if (isPartitionTable)
+	{
+		ConnectAndExecuteSPIQuery(attachToParentCommand, SPI_OK_UTILITY);
+	}
+
+	if (con->cascadeToColocated == CASCADE_TO_COLOCATED_YES)
+	{
+		Oid colocatedTableId = InvalidOid;
+
+		/* For now we only support cascade to colocation for alter_distributed_table UDF */
+		Assert(con->conversionType == ALTER_DISTRIBUTED_TABLE);
+		foreach_oid(colocatedTableId, con->colocatedTableList)
+		{
+			if (colocatedTableId == con->relationId)
+			{
+				continue;
+			}
+			char *qualifiedRelationName = quote_qualified_identifier(con->schemaName,
+																	 con->relationName);
+
+			TableConversionParameters cascadeParam = {
+				.relationId = colocatedTableId,
+				.shardCountIsNull = con->shardCountIsNull,
+				.shardCount = con->shardCount,
+				.colocateWith = qualifiedRelationName,
+				.cascadeToColocated = CASCADE_TO_COLOCATED_NO_ALREADY_CASCADED
+			};
+			TableConversionReturn *colocatedReturn = con->function(&cascadeParam);
+			foreignKeyCommands = list_concat(foreignKeyCommands,
+											 colocatedReturn->foreignKeyCommands);
+		}
+	}
+
+	/* recreate foreign keys */
+	TableConversionReturn *ret = NULL;
+	if (con->conversionType == ALTER_DISTRIBUTED_TABLE)
+	{
+		if (con->cascadeToColocated != CASCADE_TO_COLOCATED_NO_ALREADY_CASCADED)
+		{
+			char *foreignKeyCommand = NULL;
+			foreach_ptr(foreignKeyCommand, foreignKeyCommands)
+			{
+				ConnectAndExecuteSPIQuery(foreignKeyCommand, SPI_OK_UTILITY);
+			}
+		}
+		else
+		{
+			ret = palloc0(sizeof(TableConversionReturn));
+			ret->foreignKeyCommands = foreignKeyCommands;
+		}
+	}
+
+	return ret;
+}
+
+
+/*
+ * EnsureTableNotReferencing checks if the table has a reference to another
+ * table and errors if it is.
+ */
+void
+EnsureTableNotReferencing(Oid relationId)
+{
+	if (TableReferencing(relationId))
+	{
+		ereport(ERROR, (errmsg("cannot complete operation "
+							   "because table has a foreign key")));
+	}
+}
+
+
+/*
+ * EnsureTableNotReferenced checks if the table is referenced by another
+ * table and errors if it is.
+ */
+void
+EnsureTableNotReferenced(Oid relationId)
+{
+	if (TableReferenced(relationId))
+	{
+		ereport(ERROR, (errmsg("cannot complete operation "
+							   "because a foreign key references to table")));
```

Comment:
because the table is referenced by a foreign key

maybe also nice to include the table name (using get_rel_name). That's useful because alter_table_set_access_method could be used in a query on pg_class, and then it's obvious which table was the problem.